### PR TITLE
Feature/investigate different sqlite library

### PR DESCRIPTION
**Pull request by achocron**
Changed SQLite library to SQLite-net-pcl for compatibility

### CODE REVIEW

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Microsoft.Azure.Mobile.Windows.Shared.projitems [hunk 1]
```diff
@@ -20,6 +20,8 @@
     <Compile Include="$(MSBuildThisFileDirectory)Channel\IChannelUnit.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Ingestion\Http\HttpIngestionException.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Ingestion\Http\IHttpNetworkAdapter.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Storage\IStorageAdapter.cs" />
```
Comment (dhei) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Microsoft.Azure.Mobile.Windows.Shared.projitems]: You can undo change in this file as it's not needed.

File: Tests/Microsoft.Azure.Mobile.Test.Windows/Storage/FakeStorageTest.cs [hunk 2]
```diff
@@ -56,19 +74,81 @@ public void ShutdownSucceed()
...
+                .Throws(new StorageException());
+            var fakeStorage = new Mobile.Storage.Storage(mockAdapter.Object);
+            try
+            {
+                fakeStorage.PutLogAsync("channel_name", new TestLog()).RunNotAsync();
+            }
+            catch (Exception e)
+            {
+                Assert.IsNotNull(e as StorageException);
+            }
```
Comment (dhei) [inline Tests/Microsoft.Azure.Mobile.Test.Windows/Storage/FakeStorageTest.cs]: If fakeStorage.PutLogAsync().RunNotAsync() doesn't throw any exception, shouldn't this test fail? Can you modify your test to cover this scenario?

File: Tests/Microsoft.Azure.Mobile.Test.Windows/Storage/MockStorage.cs [hunk 3]
```diff
@@ -8,7 +8,6 @@ namespace Microsoft.Azure.Mobile.Test.Storage
 {
     public class MockStorage : IStorage
     {
-
```
Comment (dhei) [inline Tests/Microsoft.Azure.Mobile.Test.Windows/Storage/MockStorage.cs]: nitpick: you can undo this change as well :)

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs [hunk 4]
```diff
@@ -235,7 +207,7 @@ internal Storage(IStorageAdapter storageAdapter)
             }
```
Comment (dhei) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs]: Two questions:
(1) If we are throwing the same StorageException(errorMessage, e) for all four types of exceptions, can we simplify this code to check for exception type and throw the exception once?
(2) What's the reason ignoring exception outside the four types here? Should we catch and re-throw?

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs [hunk 5]
```diff
@@ -298,13 +279,13 @@ internal Storage(IStorageAdapter storageAdapter)
                 logs?.AddRange(retrievedLogs);
                 return batchId;
             }
-            catch (DbException e)
+            catch (StorageException)
             {
-                throw new StorageException("Error retrieving logs from storage", e);
+                throw new StorageException("Error retrieving logs from storage");
```
Comment (dhei) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs]: >throw new StorageException("Error retrieving logs from storage"); [](start = 16, length = 65)

Catch and re-throw the same storage exception with all internal information stripped off. Can we keep `throw new StorageException("Error retrieving logs from storage", e);` here?

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs [hunk 6]
```diff
@@ -261,31 +233,40 @@ internal Storage(IStorageAdapter storageAdapter)
         /// <exception cref="StorageException"/>
         public async Task<string> GetLogsAsync(string channelName, int limit, List<Log> logs)
         {
-            await OpenDbAsync().ConfigureAwait(false);
+            await StartTaskAsync().ConfigureAwait(false);
             logs?.Clear();
             var retrievedLogs = new List<Log>();
             MobileCenterLog.Debug(MobileCenterLog.LogTag, $"Trying to get up to {limit} logs from storage for {channelName}");
             try
```
Comment (dhei) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs]: >try [](start = 12, length = 3)

General feedback: I'm not a big fan of a big try-catch block that covers more than 40 lines of code plus multiple operations, except in test code. It makes the code less maintainable and less readable. I'd like to see this try-catch block refactored into shorter ones.

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs [hunk 7]
```diff
@@ -261,31 +233,40 @@ internal Storage(IStorageAdapter storageAdapter)
...
+                    {
+                        var log = LogSerializer.DeserializeLog(entry.Log);
+                        retrievedLogs.Add(log);
+                        idPairs.Add(Tuple.Create(log.Sid, Convert.ToInt64(entry.Id)));
+                    }
+                    catch (JsonException e)
+                    {
+                        MobileCenterLog.Error(MobileCenterLog.LogTag, "Cannot deserialize a log in storage", e);
+                        failedToDeserializeALog = true;
+                        await DeleteLogAsync(channelName, entry.Id).ConfigureAwait(false);
```
Comment (dhei) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs]: >await DeleteLogAsync(channelName, entry.Id).ConfigureAwait(false); [](start = 24, length = 66)

Question: is it better to throw or ignore storage exception from DeleteLogAsync? What's your reasoning of throwing storage exception here?

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/StorageAdapter.cs [hunk 8]
```diff
@@ -1,113 +1,93 @@
...
-                }
+                var table = _dbConnection.Table<T>();
+                return await table.Where(pred).Take(limit).ToListAsync().ConfigureAwait(false);
+            }
+            catch (SQLiteException e)
+            {
+                throw new StorageException(e);
             }
-            return rows;
-        }
```
Comment (dhei) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/StorageAdapter.cs]: nitpick: remove extra empty line

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/StorageAdapter.cs [hunk 9]
```diff
@@ -1,113 +1,93 @@
...
-        /// <summary>
-        /// Disposes the adapter
-        /// </summary>
-        public void Dispose()
+        public async Task<int> DeleteAsync<T>(Expression<Func<T, bool>> pred) where T : new()
         {
-            _dbConnection.Dispose();
+            try
+            {
+                var numDeleted = 0;
```
Comment (dhei) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/StorageAdapter.cs]: >numDeleted [](start = 20, length = 10)

nitpick: use complete word and avoid abbreviation in naming as best practice

Comment (achocron) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/StorageAdapter.cs]: Resharper yells at me if I don't use var haha

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs (hunk 7 above)
Comment (achocron) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs]: Yeah you're right i can ignore it

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs (hunk 4 above)
Comment (achocron) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs]: In general I try to avoid catch-alls, but perhaps we can here since all this does is wrap and re throw the exception

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs (hunk 5 above)
Comment (achocron) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Storage/Storage.cs]: We can, but ultimately it makes the log much harder to read

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Microsoft.Azure.Mobile.Windows.Shared.projitems (hunk 1 above)
Comment (achocron) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Microsoft.Azure.Mobile.Windows.Shared.projitems]: Why not?

File: Tests/Microsoft.Azure.Mobile.Test.Windows/Storage/MockStorage.cs (hunk 3 above)
Comment (achocron) [inline Tests/Microsoft.Azure.Mobile.Test.Windows/Storage/MockStorage.cs]: Are you referring to the deletion of an extra newline?

File: SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Microsoft.Azure.Mobile.Windows.Shared.projitems (hunk 1 above)
Comment (dhei) [inline SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Microsoft.Azure.Mobile.Windows.Shared.projitems]: At the time of my last comment, the only change to this file is moving 'IStorageAdapter.cs' and 'StorageAdapter.cs' around, not an actual change. That's why I said you can undo this file. Now I saw you actual have some change, so you can ignore me. 😄

File: Tests/Microsoft.Azure.Mobile.Test.Windows/Storage/MockStorage.cs (hunk 3 above)
Comment (dhei) [inline Tests/Microsoft.Azure.Mobile.Test.Windows/Storage/MockStorage.cs]: Yes.
